thunder/executors/nvfuserex_impl.py

```diff
@@ -2071,8 +2071,7 @@ def copy_(
 ) -> Any:
     nvcopy_from = getnv(copy_from, fd, lc_to_nv_map)
     nvcopy_to = getnv(copy_to, fd, lc_to_nv_map)
-    alias_output = fd.ops.set(nvcopy_from)
-    fd.add_output(alias_output, alias_input=nvcopy_to)
+    fd.add_output(nvcopy_from, alias_input=nvcopy_to)
```

Comment:
Are we supposed to do this? I'm worried that this will give us wrong program.

Reply:
This PR is composed with the revert of https://github.com/Lightning-AI/lightning-thunder/pull/1209 and the test update. 
I thought `fd.ops.set` is something ideally we want to avoid as per https://github.com/Lightning-AI/lightning-thunder/issues/1173

Reply:
sorry about not being more clear in #1173 

We still wanted to have the set for `copy_`. I think the root-cause with issue #1173 is that we are returning `nvcopy_from` in the fusion region.

Reply:
why isn't there a copy op?

Reply:
There is a copy op, the difference is, whether we are returning a different buffer, or the aliased source.

I think the example below would help.

The thunder program below behaves differently, depends on what's the return value.
```
import thunder
import torch
from thunder.core import prims

def foo(x):
    x_relu = prims.abs(x)
    old_x = prims.copy_(x_relu, x)
    return old_x  # this returns x, the aliased source from the program.
    #return x_relu  # this returns the x_relu, which should be in a different buffer.

jfoo = thunder.jit(foo)

x = torch.randn(2, 4, device="cuda")
o = jfoo(x)

print(thunder.last_traces(jfoo)[-1])
print(x)
print(o)
print(x.data_ptr())
print(o.data_ptr())
```

They are translated as nvfuser programs correspondingly.
```
import torch
from nvfuser import FusionDefinition, DataType

def nvfuser_fusion_id4(fd : FusionDefinition) -> None :
    T0 = fd.define_tensor(shape=[2, 4], contiguity=[True, True], dtype=DataType.Float, is_cpu=False, stride_order=[1, 0])
    T1 = fd.ops.neg(T0)
    T2 = fd.ops.set(T1)
    fd.add_output(T2, T0)
    fd.add_output(T2)
    # fd.add_output(T1)  # doing this would cause us to return an extra buffer.
```
The part of `copy_` is translated to
```
T2 = fd.ops.set(T1)  # this is the extra copy we added.
fd.add_output(T2, T0)  # the extra copy is used to overwrite T0.
```
When we return T2 as output of the fusion, since T2 is marked as alias to T0 (sharing storage), we are not returning an extra buffer here.

However, the story changes when we  **return T1** from the fusion. Since `T1` is a new buffer in the program. You can play with this using vvv:
```
with FusionDefinition() as fd:
    nvfuser_fusion_id4(fd)

inputs = [
    torch.testing.make_tensor((2, 4), dtype=torch.float32, device='cuda:0'),
]

print(inputs)
o = fd.execute(inputs)
print(inputs)

print(o[0].data_ptr())
print(inputs[0].data_ptr())
```

You'll notice that without the extra `fd.ops.set` in `copy_`, nvfuser won't be able to faithfully represent the intended thunder behavior.

Reply:
okay, then I don't think this PR deserves a merge.

f0eac737e2a32792f2b0a4101a3075b730329b9f: compile 54s, 11.07ms
82dc7a7c0f4739604f7a32b98defd3158abf5ac9: compile 97s, 14.28ms